### PR TITLE
front-end frameworks, beginner's guide

### PR DESCRIPTION
**Pull request by hillalex**
Feel like this could do with some closing remarks, maybe linking to this other good post I read today https://stackoverflow.blog/2020/02/03/is-it-time-for-a-front-end-framework/

Perhaps @EmmaLRussell might have some edits/additions as well - feel free to go for it if so!

### CODE REVIEW

**Comment by hillalex** [comment]
nb date needs fixing before merging

File: content/blog/js-beginners-guide.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+<error-list errors="myErrorArray"></error-list>
+```
+
+A directive is like a custom data attribute on a regular HTML element, with special behaviour attached:
+
+```
+<span v-if="hasError" v-text="message"></span>
+```
+
+A component is a combination of a **template** that defines the presentation, and a script that defines
```
Comment (richfitz) [inline content/blog/js-beginners-guide.md:54]: This appears to be the second definition of "A component" - is this an elaboration of the definition above? Can this be made clearer, especially as it involves directives?

File: content/blog/js-beginners-guide.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+A directive is like a custom data attribute on a regular HTML element, with special behaviour attached:
+
+```
+<span v-if="hasError" v-text="message"></span>
+```
+
+A component is a combination of a **template** that defines the presentation, and a script that defines 
+behaviour.
+
+Here's an example of an HTML template in Vue.js. It uses the v-for directive to loop over an array:
```
Comment (richfitz) [inline content/blog/js-beginners-guide.md]: Link to vue.js (and below React)?

File: content/blog/js-beginners-guide.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+```
+<ul className="list">
+    {errors.map(e => <li>{e}</li>)}
+</ul>
+```
+
+## Application architecture
+Finally, you might have heard of [Flux](https://facebook.github.io/flux/) - it's a code architecture pattern pioneered by Facebook, the creators of React,
+ and now widely adopted for organising complex front-end applications. React by itself is just a lightweight library for writing components, but it is commonly 
+used in conjunction with [Redux](https://react-redux.js.org/) which is a Flux implementation.
```
Comment (richfitz) [inline content/blog/js-beginners-guide.md]: Add something to indicate earlier that implementations of flux exist for other frameworks than React?

File: content/blog/js-beginners-guide.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+</ul>
+```
+
+## Application architecture
+Finally, you might have heard of [Flux](https://facebook.github.io/flux/) - it's a code architecture pattern pioneered by Facebook, the creators of React,
+ and now widely adopted for organising complex front-end applications. React by itself is just a lightweight library for writing components, but it is commonly 
+used in conjunction with [Redux](https://react-redux.js.org/) which is a Flux implementation. 
+
+The best way to understand Flux is probably to read the [docs](https://facebook.github.io/flux/docs/overview) over at Facebook. 
+But in a nutshell, it's a uni-directional data flow pattern inspired by functional programming.
```
Comment (richfitz) [inline content/blog/js-beginners-guide.md]: Can you expand the nutshell a bit? "Uni-directional" seems to be the important word here but I don't understand why that is important in this context

File: content/blog/js-beginners-guide.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+It can add a lot of boilerplate code to your app, so it's not always the right solution. If you're just writing a handful 
+of mostly independent components, you almost certainly don't need it. If you have a complex, dynamic app with many components
+ that need to share state, then it can be really useful.
+
+The Redux homepage suggests:
```
Comment (richfitz) [inline content/blog/js-beginners-guide.md]: Link to the suggestion

**Review by richfitz — COMMENTED** [review]
This is nice and very accessible. A few links to tutorials/etc that you found particularly useful could be worthwhile. Otherwise a handful of comments below

File: content/blog/js-beginners-guide.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+```
+
+A directive is like a custom data attribute on a regular HTML element, with special behaviour attached:
+
+```
+<span v-if="hasError" v-text="message"></span>
+```
+
+A component is a combination of a **template** that defines the presentation, and a script that defines 
+behaviour.
```
Comment (EmmaLRussell) [inline content/blog/js-beginners-guide.md:55]: "behaviour and data"? Could include a really simple script which corresponds to your error list template to show what that might look like?

File: content/blog/js-beginners-guide.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+```
+
+## Application architecture
+Finally, you might have heard of [Flux](https://facebook.github.io/flux/) - it's a code architecture pattern pioneered by Facebook, the creators of React,
+ and now widely adopted for organising complex front-end applications. React by itself is just a lightweight library for writing components, but it is commonly 
+used in conjunction with [Redux](https://react-redux.js.org/) which is a Flux implementation. 
+
+The best way to understand Flux is probably to read the [docs](https://facebook.github.io/flux/docs/overview) over at Facebook. 
+But in a nutshell, it's a uni-directional data flow pattern inspired by functional programming.
+
```
Comment (EmmaLRussell) [inline content/blog/js-beginners-guide.md:86]: Might be useful to include your diagram from the poster here too, which is a really nice encapsulation of Flux.

**Review by EmmaLRussell — COMMENTED** [review]
Looks great! Might be nice to kick it off with a quick outline of why frameworks are useful (corresponding to the top part of the poster) - which would be a good place to link to that stackoverflow blog post which goes into that in a lot more detail. 

Maybe something like:
 "If you're creating an interactive web application based on Javascript, you'll need to:
- manage changing state
- organise a potenially complex system of code components reflecting and updaing that state
-  dynamically update HTML
- write a suite of tests to check all this works

This can quickly become unmanageable. A Javascript framework could be useful because it provides a ready-made structure you can use to organise your application, including:
- state management and data binding
- component based architecture
- HTML templating
- testing utilities

Error prone 'plumbing' code is eliminated and you can concentrate on the content of your application."

**Pull request merged**